### ISSUE TITLE
Association ExpressionPredicate

### ISSUE THREAD

**Issue by tb-mtg** [linq2db/linq2db#841]
I am trying to use the Association ExpressionPredicate (with constants) on an old legacy database to join a products table to a horrible lookup table and get the following exception:

**Exception message:** 
'ConvertToCaseCompareTo(y.CompanyCode.ToUpper(), x.CompanyCode.ToUpper()).Value' cannot be converted to SQL.

```
**Stack trace:**
  [LinqException: 'ConvertToCaseCompareTo(y.CompanyCode.ToUpper(), x.CompanyCode.ToUpper()).Value' cannot be converted to SQL.]
   LinqToDB.Linq.Builder.ExpressionBuilder.ConvertToSql(IBuildContext context, Expression expression, Boolean unwrap) in c:\projects\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:1016
```

**Step to Reproduce**
```
// usage user input variables
var uiCompanyCode = "S1";
var uiLookupTypeCode = "PGMJ";
var uiproductCodes = {"02071","00333","00418"};
var uiMajorGroupCodes = {"080",	"090"};

// usage (both throw same error)
var list1 = (from x in db.GetTable<Product>
             where x.CompanyCode == uiCompanyCode &&
                   uiproductCodes.Contains(x.ProductCode)
             select x.ProductCode, 
                    Product = x.Description, 
                    x.MajorGroupCode, 
                    MajorGroup = x.MajorGroup.Description
            ).ToList();

var list2 = (from x in db.GetTable<HorribleLookupTable>
             where x.CompanyCode == uiCompanyCode &&
                   x.LookupTypeCode == uiLookupTypeCode && 
                   uiMajorGroupCodes.Contains(x.LookupCode)
             select x.LookupCode, x.Description, x.ProductList.Count
            ).ToList();
```
**Mapping classes**
```
[Table("INP35")]
public class Product{
	[Column("CONO35")]	public string CompanyCode { get; set; }
	[Column("PNUM35")]	public string ProductCode { get; set; }
	[Column("PDES35")]	public string Description { get; set; }
	[Column("PGMJ35")]	public string MajorGroupCode { get; set; }

	[Association(CanBeNull = true, ExpressionPredicate = NameOf(Predicates_MajorGroup))]	
public HorribleLookupTable MajorGroup { get; set; }

	public static Expression<Func<Product, HorribleLookupTable, bool>> Predicates_MajorGroup { get; set; }
}

[Table("INP15")]
public class HorribleLookupTable{
	[Column("CONO15")]	public string CompanyCode { get; set; }
	[Column("PRMT15")]	public string LookupTypeCode { get; set; }
	[Column("PSAR15")]	public string LookupCode { get; set; }
	[Column("PRMD15")]	public string Description { get; set; }

	[Association(CanBeNull = true, ExpressionPredicate = NameOf(Predicates_MajorGroup))]
	public IEnumerable<Product> ProductList { get; set; }

	public static Expression<Func<Product, HorribleLookupTable, bool>> Predicates_MajorGroup { get; set; }
}

public class Predicates{
	public static Expression<Func<Product, HorribleLookupTable, bool>> MajorGroup = (x, y) => y.CompanyCode == x.CompanyCode && y.LookupTypeCode == "PGMJ" && x.MajorGroupCode == y.LookupCode;
}
```

### Environment details
linq2db version: *1.9.0*  
Database Providers: Tested on both **SqlServer** & *[Linq2DB4iSeries](https://github.com/LinqToDB4iSeries/Linq2DB4iSeries)* 
Operating system: *Windows 10*
Framework version: *.NET Framework 4.6.2*

**Comment by sdanyliv** [comment]
Which provider do you use?
Also i can not see assignment it is dynamical?
```
public static Expression<Func<Product, HorribleLookupTable, bool>> Predicates_MajorGroup { get; set; }
``` 

Expression for `HorribleLookupTable` must be be the following type:
```cs
Expression<Func<HorribleLookupTable, Product, bool>>
```
First parameter LEFT table, then RIGHT and result value

**Comment by tb-mtg** [comment]
I am using the [Linq2DB4iSeries](https://github.com/LinqToDB4iSeries/Linq2DB4iSeries) provider.

I have changed the Mapping classes to below and still get same error.
```
[Table("INP35")]
public class Product {
  [Column("CONO35")] public string CompanyCode { get; set; }
  [Column("PNUM35")] public string ProductCode { get; set; }
  [Column("PDES35")] public string Description { get; set; }
  [Column("PGMJ35")] public string MajorGroupCode { get; set; }

  [Association(CanBeNull = true, ExpressionPredicate = nameof(Predicates_MajorGroup))]
  public HorribleLookupTable MajorGroup { get; set; }

  static Expression<Func<Product, HorribleLookupTable, bool>> Predicates_MajorGroup = (p, h) =>
    p.CompanyCode == h.CompanyCode && h.LookupTypeCode == "PGMJ" && p.MajorGroupCode == h.LookupCode;
}

[Table("INP15")]
public class HorribleLookupTable {
  [Column("CONO15")] public string CompanyCode { get; set; }
  [Column("PRMT15")] public string LookupTypeCode { get; set; }
  [Column("PSAR15")] public string LookupCode { get; set; }
  [Column("PRMD15")] public string Description { get; set; }

  [Association(CanBeNull = true, ExpressionPredicate = nameof(Predicates_MajorGroup))]
  public IEnumerable<Product> ProductList { get; set; }

  static Expression<Func<HorribleLookupTable, Product, bool>> Predicates_MajorGroup = (h, p) =>
    p.CompanyCode == h.CompanyCode && h.LookupTypeCode == "PGMJ" && p.MajorGroupCode == h.LookupCode;
}
```

**Comment by sdanyliv** [comment]
@tb-mtg, it's provider specific bug, i'll try to reproduce.

**Comment by tb-mtg** [comment]
I might be wrong but I do not think it is a provider specific bug as I have also tried it with the ***SqlServer*** provider and still got and the same exception.

**Comment by sdanyliv** [comment]
I think I've fixed your problem. Looks like there is some option for string comparison using function Sql.ConvertToCaseCompareTo which was not properly handled.
Please verify. 
Thanks.

**Comment by tb-mtg** [comment]
Thanks. Do I need to verify or was that meant for somebody else?

**Comment by sdanyliv** [comment]
@tb-mtg, please verify that my fix works for you. I've created synthetic test and it may fail in your situation.
It is not merged yet, but you can grab source from https://github.com/linq2db/linq2db/tree/issue_841 - issue_841 branch

**Comment by tb-mtg** [comment]
@sdanyliv, yes i have verified your code and it did fix the problem. Thanks

**Comment by sdanyliv** [comment]
Fixed in #842
